onedal/datatypes/dlpack/data_conversion.cpp

```diff
@@ -0,0 +1,183 @@
+#endif // ONEDAL_DATA_PARALLEL
+
+namespace oneapi::dal::python::dlpack {
+
+template <typename T, typename managed_t>
+inline dal::homogen_table convert_to_homogen_impl(managed_t* dlm_tensor,
+                                                  bool readonly,
+                                                  py::object q_obj) {
+    dal::homogen_table res{};
+
+    DLTensor tensor = dlm_tensor->dl_tensor;
+
+    // if a nullptr, return an empty.
+    if (!tensor.data) {
+        return res;
+    }
+
+    // Get pointer to the data following dlpack.h conventions.
+    // use static cast because data is known to be void*, and reintrepret_cast
+    // since we know that T* is of a set specific types that we will guarantee
+    // for the compiler.
+    const auto* const ptr =
+        reinterpret_cast<const T*>(static_cast<char*>(tensor.data) + tensor.byte_offset);
+
+    // get shape, if 1 dimensional, force col count to 1
+    std::int64_t row_count, col_count;
+    row_count = tensor.shape[0];
+    col_count = get_ndim(tensor) == 1 ? 1l : tensor.shape[1];
+
+    // get data layout for homogeneous check
+    const dal::data_layout layout = get_dlpack_layout(tensor);
+
+    // create the dlpack deleter, which requires calling the deleter from the managed DL tensor.
+    // This must be done instead of a decref, as the data doesn't necessarily come from python,
+    // It is expected that deleter handles memory cleanup (including possible python decrefs)
+    const auto deleter = [dlm_tensor](const T* data) {
+        if (dlm_tensor->deleter != nullptr) {
+            dlm_tensor->deleter(dlm_tensor);
+        }
+    };
+
+    // check_dlpack_oneAPI_device will check if data is on a oneAPI device. If it is on an
+    // unsupported device it will throw an error.
+    if (check_dlpack_oneAPI_device(tensor.device.device_type)) {
+#ifdef ONEDAL_DATA_PARALLEL
+        // if located on a SYCL device, use the queue.
+#define MAKE_QUEUED_HOMOGEN(pointer)                     \
+    res = dal::homogen_table(queue,                      \
+                             pointer,                    \
+                             row_count,                  \
+                             col_count,                  \
+                             deleter,                    \
+                             std::vector<sycl::event>{}, \
+                             layout);
+        // generate queue from dlpack device information
+        // If a queue is given, it will override the general queue that would be generated
+        // from get_queue_by_device_id. Note, this behavior only occurs if on a oneAPI device.
+        sycl::queue queue;
+        queue = !q_obj.is(py::none()) ? get_queue_from_python(q_obj)
+                                      : get_queue_by_device_id(tensor.device.device_id);
+
+        if (readonly) {
+            MAKE_QUEUED_HOMOGEN(ptr);
+        }
+        else {
+            auto* const mut_ptr = const_cast<T*>(ptr);
```

Comment:
```suggestion
            auto* mut_ptr = const_cast<T*>(ptr);
```

Same thought about making the pointer `const` as above.
By the way, the data is provided non-`const`, you are making it `const` by choice above and `const`-casting it away again here.
At first glance, the non-`const` usage is dominant and I don't see a point in ever making the data `const`. Did I miss something?

Reply:
ahh, I lifted this part from here, maybe we need to modify it there too? https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/onedal/datatypes/sycl_usm/data_conversion.cpp#L113

Reply:
Just as an update: with the discussion of the readonly, I removed the aspect entirely.